### PR TITLE
DIG-776: Federating Logistic Regression Classifier

### PR DESCRIPTION
**Pull request by AliRZ-02**
## Description
This PR deals with the [DIG-776](https://candig.atlassian.net/browse/DIG-776) ticket, a child of the [DIG-774](https://candig.atlassian.net/browse/DIG-774) ticket. The latter is a story focusing on implementing the new federated design created in [DIG-757](https://candig.atlassian.net/browse/DIG-757) into the federated-learning repo. The former focuses on choosing a model that is one of the top-performing models from earlier exploratory data analysis, while also upgrading the preprocessing steps in the model. The upgrades have been tested for federated systems containing 2 and 3 clients.

## Changelog
- Implemented a Logistic Regression Model Pipeline to allow for a federated version of the Synthea Dataset Analysis
- Implemented a quick start script to use while setting up the federation service
- Upgraded Pre-Processing Pipeline based on earlier EDA
- Upgraded Docker-Compose Configuration scripts to fit with new pipeline
- Modified Erroring Filepaths
- Made fl-client and fl-server Docker Containers lighter 

## Dependencies
- N/A

## Testing
- You can test the experimental setup by ensuring you have a copy of the 10yrs Synthea mCODE data installed on your machine, and then using the quickstart script to set up instances of the services needed.

Eg. Let's say the 10yrs directory is in the root federated-learning folder. Then from the root federated-learning folder, perform:
```bash
./experiments/synthea-breast-cancer/winter2022/Federated/quickstart.sh -i -s 10yrs/female 3 5000
```

This will generate 3 clients and 1 server, alongside ingesting the `10yrs/female` data into an instance of Katsu, accessible through the GraphQL-interface generated. This will take quite a long time, up to 35 minutes if ingesting the data as well. For more information on the flags and arguments, refer to the help display of the quickstart script (accessed with `-h` flag).

## Notes
- The changes in this PR are slightly suspect at times (eg multiple copies of utils.py, helpers, etc.) but this was an intentional decision since we wanted to leave much of the refactoring work to future tickets.
- A corpus splitting function was added to the quick start script that is enabled by leaving off the `-s` flag. It was simply done so because it was easy to implement while implementing the other parts of the data ingest, though it won't be used until the [DIG-782](https://candig.atlassian.net/browse/DIG-782) ticket.
- Upon implementation with Flower, the accuracy of the Logistic Regression model decreased. This is unexpected as the dataset/pre-processing/classification are the same. Some optimisation was done to improve the accuracy of the classifier, such as changing solvers, having an equal number data for each stage and increasing the number of iterations. While this did help, it still did not reach the accuracy of the non-federated classifier.

### CODE REVIEW

**Review by yuriy-pavlov — CHANGES_REQUESTED** [review]
Great work! Just a couple small fixes needed, then you can go ahead and merge.

`services/fl-server/init.sh` should be renamed to `entrypoint.sh` or something similar.`init = initialize` but there's no initialization happening in this script, it's just sleeping and then launching an app.

File: experiments/synthea-breast-cancer/winter2022/Federated/server.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    """Send round number to client."""
+    return {"rnd": rnd}
+
+
+def get_eval_fn(model: LogisticRegression):
+    """Return an evaluation function for server-side evaluation."""
+
+    print("getting test data")
+    # Load test data here to avoid the overhead of doing it in `evaluate` itself
+    _, (X_test, y_test) = utils.load_data()
```
[the remark this answers is not in the document]
Comment (AliRZ-02) [reply experiments/synthea-breast-cancer/winter2022/Federated/server.py:22]: `(X_train, y_train)`, basically just a tuple containing the training data, given that we simply want to use the testing data for the evaluation function.

File: experiments/synthea-breast-cancer/winter2022/Federated/utils.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+        for _, row in df.iterrows():
+            X.append([row.surgical, row.radiation, row.cancerStatus,
+                     row.diagnosisAge, row.primary, row.nodes, row.numberOfMeds])
+            y.append(row.stage)
+
+        X_train, X_test, y_train, y_test = train_test_split(
+            X, y, test_size=0.2, random_state=RANDOM_STATE)
+        return scale_data(((X_train, y_train), (X_test, y_test)))
+
+    def undersample_majority_class(df: DataFrame) -> DataFrame:
```
[the remark this answers is not in the document]
Comment (AliRZ-02) [reply experiments/synthea-breast-cancer/winter2022/Federated/utils.py:163]: We ended up finding a use for it while working on later PRs so hence why we decided to keep it for now but noted nonetheless. And that's a good idea, we may in fact make it abstract in the refactoring process.
